Introduced hp::Refinement::p_adaptivity_fixed_number().

marcfehling opened the pull request:
To compare the impact of different p-adaptation strategies, I was looking for a restriction on all of them that made them adapt a predefined number of cells only. For this, we need a strategy for p-adaptation comparable to `GridRefinement::refine_and_coarsen_fixed_fraction()`. This PR provides such a strategy.

Since most of the functionality is similar to pre-existing code, I wanted to re-use some helper functions from `parallel::distributed::GridRefinement` namespace. The first commit basically moves the corresponding functions from an anonymous namespace to a globally accessible one. Please leave a note if you do not agree to this procedure.

The second commit provides the new functionality.

bangerth commented on include/deal.II/hp/refinement.h:
```diff
@@ -254,6 +254,49 @@ namespace hp
+     * cell. Parameters @p p_refine_fraction and @p p_coarsen_fraction need to be
+     * in the interval $[0,1]$.
+     *
+     * @note Preceeding calls of Triangulation::prepare_for_coarsening_and_refinement()
```
Do you mean "Following calls" or "Later calls"?

Maybe suggest to call that function first. Or does the current function clear refine/coarsen flags on cells where we will use p-adaptivity, and consequently one *has* to call `prepare_c_and_r` again at a later time (maybe through `execute_c_and_r`)?

marcfehling replied:
This was meant as a warning that `prepare_c_and_r` will change `refine` and `coarsen` flags which is the decision basis for this function.

So preceeding or earlier calls of `prepare_c_and_r` might change the output of this function. Maybe I should be more specific like "Don't call `prepare_c_and_r` before this function unless you know what you are doing."

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+
+      // 1.) First extract from the vector of indicators the ones that
+      //     correspond to cells that we locally own.
+      unsigned int   n_flags_refinement(0), n_flags_coarsening(0);
```
Put these on separate lines and initialize via `=0` instead of `(0)` for common style with many other places

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+                criteria(cell->active_cell_index());
+          }
+      indicators_refinement.grow_or_shrink(n_flags_refinement);
+      indicators_coarsening.grow_or_shrink(n_flags_coarsening);
```
ha, I didn't remember we had that function :-)

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+      //
+      //     For cases in which no or all cells flagged for refinement and/or
+      //     coarsening are subject to p-adaptation, we usually pick thresholds
+      //     that appeal to all or none of the cells at once. However here, we
```
appeal -> apply?

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+      // 4.) Perform p-adaptation with absolute thresholds.
+      Number threshold_refinement(0.), threshold_coarsening(0.);
+      auto   reference_compare_refine  = std::cref(compare_refine);
+      auto   reference_compare_coarsen = std::cref(compare_coarsen);
```
It would be easier to read if you just inlined these variable's values wherever they are used.

marcfehling replied:
I'm not sure what you mean by that. I need to place all of these variables in front of the conditional statement to keep them in the correct scope, so I'm not sure if there is any benefit if I move them somewhere else. Or should I be more elaborate in the comments on why we need these reference wrappers here?

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+      auto   reference_compare_coarsen = std::cref(compare_coarsen);
+      if (const parallel::distributed::Triangulation<dim, spacedim>
+            *parallel_tria = dynamic_cast<
+              const parallel::distributed::Triangulation<dim, spacedim> *>(
```
shouldn't this also apply to the fully distributed triangulation?

marcfehling replied:
Yes, it's the same situation that led me to the decision above. I will apply the same changes here as well.

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+          if (target_index_refinement == 0)
+            reference_compare_refine = std::cref(compare_false<Number>);
+          else if (target_index_refinement == n_global_flags_refinement)
+            reference_compare_refine = std::cref(compare_true<Number>);
```
ah, I see how this can be used!

bangerth commented on source/hp/refinement.cc:
```diff
@@ -207,6 +228,182 @@ namespace hp
+          // serial implementation (and parallel::shared implementation)
+          //
+
+          // 2.) Determine amount of cells scheduled for p-adaptation.
```
amount -> number

bangerth commented on tests/hp/p_adaptivity_flags.cc:
```diff
@@ -181,6 +181,28 @@ test()
 
+  deallog << "p-adaptivity fixed number" << std::endl;
+  {
+    Triangulation<dim>  tria;
```
Just put it into a separate test. this makes it easier to find bugs.

marcfehling replied:
I see. While there, I will divide this particular test into multiple ones, each checking one particular component.

bangerth commented:
/rebuild

marcfehling commented on source/hp/refinement.cc:
```diff
@@ -168,8 +188,9 @@ namespace hp
-      if (const parallel::TriangulationBase<dim, spacedim> *parallel_tria =
-            dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+      if (const parallel::distributed::Triangulation<dim, spacedim>
+            *parallel_tria = dynamic_cast<
```
I noticed that we do not need a the succeeding communication step for the `parallel::shared::Triangulation` class, since we will use distributed vectors in that case (similar to `step-17`).

However, I totally forgot about the new `parallel::fullydistributed::Triangulation` class. I'll revert back to the general class, and add another check whether it's a `parallel::shared::Triangulation`.

marcfehling commented:
I was unfocused and had some kind of accident during the last rebasing session... I fixed everything now and addressed all suggestions. Waiting until testers finish, but feel free to give it another look!

marcfehling commented:
Part of #7515.
